fake_spectra/index_table.cpp

```diff
@@ -147,3 +148,88 @@ std::valarray< std::map<int, double> > IndexTable::get_near_particles(const floa
     }
     return line_near_parts;
 }
+
+double * IndexTable::assign_cells(const int line_i, const std::valarray< std::map<int, double> > nearby_array, const float pos[])
+{
+    /***********************************/
+    printf("assigning parts of line %d to cells...\n", line_i);
+    int Ncells = 0;
+    for(std::map<int, double>::const_iterator it = nearby_array[line_i].begin(); it != nearby_array[line_i].end(); ++it) Ncells++;
+    printf("Ncells=%d\n", Ncells);
+    double * arr2 = new double [2*Ncells];
+    // initialize
+    for(int i = 0; i < Ncells; ++i)
+    {
+        arr2[2*i] = -100;
+        arr2[2*i+1] = -100;
+    }
+    //printf("arr2[0]=%f, arr2[1]=%f\n", arr2[0], arr2[1]);
+
+    // divide each sightline into an array. grid size = RESO ckpc/h
+    int N = int(boxsize/RESO);
+    double reso = boxsize/N;
+
+    int axis_i = axis[line_i];
+    double xp;
+    double yp = cofm[3*line_i+axis_i%3], zp = cofm[3*line_i+(axis_i+1)%3];
+    double dx, dy, dz;
+    // loop over all grid points along the sightline
+    for(int i = 0; i < N; ++i){
+        // the default choice is the sightline points in the x direction
+        xp = (i+0.5)*reso;
+        // find the particle index that this point along the sightline belongs to
+        double dist, min_dist = boxsize;
+        int min_ind = 0, ind = 0;
+        for(std::map<int, double>::const_iterator it = nearby_array[line_i].begin(); it != nearby_array[line_i].end(); ++it){
+            const int ipart = it->first;
+            // take into account periodicity
+            dx = abs(pos[3*ipart+axis_i-1]-xp);
+            if(dx > boxsize/2.) dx = boxsize - dx;
+            dy = abs(pos[3*ipart+axis_i%3]-yp);
+            if(dy > boxsize/2.) dy = boxsize - dy;
+            dz = abs(pos[3*ipart+(axis_i+1)%3]-zp);
+            if(dz > boxsize/2.) dz = boxsize - dz;
+            dist = sqrt(dx*dx + dy*dy + dz*dz);
+            if(dist < min_dist){
+                min_dist = dist;
+                min_ind = ind;
+            }
+            ind++;
+        }
+        // assign to arr2
+        //printf("i=%d min_ind=%d, before: left=%f, right=%f\n", i, min_ind, arr2[2*min_ind], arr2[2*min_ind+1]);
+        if(arr2[2*min_ind] < 0)
+        {
+            arr2[2*min_ind] = xp;
+            arr2[2*min_ind+1] = xp;
+        }
+        else
+        {
+            if(xp - arr2[2*min_ind+1] < 1.01*reso)
+                arr2[2*min_ind+1] = xp;
+            else
+            {
+                if(arr2[2*min_ind] < reso && xp > boxsize/2.+0.5*reso)
+                {
+                    arr2[2*min_ind] = xp;
+                    arr2[2*min_ind+1] += boxsize;
+                    break;
+                }
+                else
```

Comment:
ok I guess there is a misunderstanding here. So I loop over points that are separated by 1 ckpc/h on the sightline, and these are **not** the pixels which have 1 km/s in width by default. For each of those points I find the mesh generating point of the minimum distance. The reason for doing this is to determine from which position to which position on a sightline is enclosed by a certain gas cell.
It's just that I find it easier to determine arr2 in this way because I don't know the exact boundaries of the voronoi mesh.

Reply:
I see. I'll try 0.5 kpc/h and see what happens. But there are two things I don't quite understand:
1. if the forest is sensitive to gas with overdensities of ~0.1-10, then the smallest cell size in this density range is 20 ckpc/h. I'm not sure how much the densest cells contribute to the spectrum though. Especially at redshift ~5, gas with overdensities of a few already have pretty saturated absorption.
2. the comoving softening length in the simulations is 0.5 kpc/h, so I guess 1 kpc/h is small enough?

This is the cell size distribution in my simulations. The cell sizes are simply calculated by (Mass/Volume)^1/3.
<img width="575" alt="screen shot 2019-02-01 at 1 40 20 pm" src="https://user-images.githubusercontent.com/30164085/52143435-7205ad00-2629-11e9-8765-d1b06f1b3389.png">